app/src/main/java/com/keylesspalace/tusky/components/preference/PreferencesFragment.kt

```diff
@@ -51,6 +51,26 @@ class PreferencesFragment : PreferenceFragmentCompat(), Injectable {
     private val iconSize by lazy { resources.getDimensionPixelSize(R.dimen.preference_icon_size) }
     private var httpProxyPref: Preference? = null
 
+    enum class ReadingOrder {
+        /** User scrolls up, reading statuses oldest to newest. Default behaviour. */
+        OLDEST_FIRST,
```

Comment:
NEWEST_FIRST is the current behavior, right? Can we have it as the default please?

Reply:
I think NEWEST_FIRST is a bug, that this fixes.

From what I can tell, up to and including v16 beta 1 Tusky  OLDEST_FIRST behaviour. I verified this by checking out 31da851f, adding the DeveloperTools functionality, and experimenting with creating "Load more" gaps of different sizes and seeing how the app behaved when I clicked the button. It kept the position in the list.

I think the NEWEST_FIRST behaviour was introduced in 6d4f5ad0279e5ff51f143742fca060dd04e1c30c with the migration to Paging3.

I can't see anything that suggests that this was a deliberate change, it's just how Paging3 works by default.

This change rapidly started annoying users, and they submitted issues about it. The earliest issue I can see is https://github.com/tuskyapp/Tusky/issues/2423 from Apr 7th.

I can't find any issues from before the change where users were asking for "Load more" to have its current behaviour. Only issues from after the change with users asking for the behaviour to be reverted.

This is why I think the best thing to do for the users is for the default to be the previous behaviour, which fixes the bug introduced in 6d4f5ad.

Reply:
But OLDEST_FIRST is loading the oldest statuses in the gap, we never did that?

Reply:
If you don't do that (i.e,. use `since_id`), with this you can get the experience where you click "Load more", the statuses load, but they don't reach the bottom of the gap (i.e., the gap is more than LOAD_AT_ONCE statuses in size).

So then there's a new "Load more" button, and you have to click again. And potentially again.

This code, using `min_id` (and with `OLDEST_FIRST`) means the UX is to click "Load more", then scroll up and read the loaded statuses, eventually encounter another "Load more", click that, and repeat, which seems like the right way to do it (and is requested from the filed issues).

Reply:
Using `min_id` is also consistent with Mastodon's paging in the `link` headers in HTTP responses.

Reply:
The default usage behavior is to read from newest to oldest - that's why the timeline is in *reverse* chronological order.
In that case, we want to load the *newest* statuses in the gap by default, so that the user who reads from newest to oldest gets an uninterrupted stream of posts. It's possible that there was a bug or otherwise incorrect behavior here in the past, but I believe that today's behavior is correct for this use case

I get that we want to have better support for the vocal minority of users who want to read every post on their timelines from oldest to newest, but we should absolutely not change such a fundamental default behavior on a whim or due to selection bias.

Reply:
As a side note, I have felt like I have maybe been coming across as adversarial in some of these discussions, and I've been trying to limit my participation for that reason. *However,* I also feel like there's an abnormally high amount of talking past each other going on, and I don't understand whether it's because one or more of us are fundamentally misunderstanding what the others are saying, or there's some hidden assumption, or something else.

Reply:
If it helps, I haven't been interpreting your feedback as adversarial. 

To address what I believe are your key points:

## "The default usage behavior is to read from newest to oldest"

That's not always true, the app is inconsistent about this behaviour.

For example:

1. Check out the current tip of `develop` (eb02d0b at the time I'm writing this)
2. Launch on a device or emulator
3. Open the Federated timeline (just because it changes very quickly and you don't have to wait long)
4. Scroll to the very top, and swipe-to-refresh to make sure you have the most recent statuses
5. Wait maybe 10 seconds, so that new statuses are available
6. Swipe-to-refresh to load new statuses

You should observe that your position in the timeline is retained, and the new statuses are loaded above the one you are currently looking at. This is oldest-first behaviour.

This behaviour is the same on at least the Federated, Local, and Home timelines.

If "the default usage behaviour is to read from newest to oldest" was correct, refreshing any of these timelines should take you to the newest of the freshly loaded statuses.

I suspect this is the source of some of the demand for this behaviour from users. They observe that, in some situations the app behaves the way that they expect, but in others it doesn't.

This also gets weird if there were more than LOAD_AT_ONCE statuses loaded and there's also a "Load more" placeholder. Because if the user clicks that then the statuses are loaded below their current position.

So swipe = load above and keep my position, "Load more" = load below and lose my position

## "vocal minority"

I agree about "vocal". There are at least 14 issues that reference this, many of which either have comments from other users saying they also want the oldest-first behaviour, or heart / thumbs-up emoji from people who haven't commented but also indicating that they want the oldest-first behaviour.

I don't think the assertion that it's a minority is knowable.

Creating a Github issue or commenting on it has a fairly high barrier to entry -- you have to know where to file the issue, you have to create a Github account if you don't already have one, and you have to take the time to actually write the issue.

Participation inequality is a thing (https://www.nngroup.com/articles/participation-inequality/), and I think it's reasonable to assume that for every 1 user who has taken the time to report this issue, there are N other users who either have the same problem, or have abandoned using the app in favour of a different one.

Unfortunately, we don't know what N is. Could be 10, could be 100, could be 1,000, or more, or less.

This also extends to Play reviews. It doesn't appear to be possible to link to individual reviews, so, from the last month or so:

"Tusky experience (if you try to read from oldest to newest): Scrolling up. Tap "Load more". Scroll down to find where you stopped before it loaded newest posts and lost scrolling position. Tap "Load more" again, rushing to scroll down so that you don't lose scrolling progress. "Load more". "Load more". Repeat 15 times [...]" 2 stars, "3 people found this review helpful"

"Looks pretty good. Nice and clean view. However the load more button defaults to going to the top of your feed losing position. Can this be switched or an option included?", 3 stars, "4 people found this review helpful"

"This App is okay for the most cases. But the main feature is complety insane. The timeline is always at the latest twee...äh..toot(?). You can't continue reading the toots from the place were you stopped reading. You always have to start at the newest toot and must search were you left", 2 stars, "25 people found this review helpful"

My intuition from all this is that N is a large number, and that changing the default approach to oldest-first is a net-win for the majority of the userbase.

You, reasonably, disagree.

We can't both be right.

## Proposal

The Tusky Mastodon account has about 15K followers.

Create a poll, run it for 7 days, and boost it once per day for those 7 days. Suggested text:

```
We are thinking of changing the default "Load more" behaviour (https://github.com/tuskyapp/Tusky/pull/3000)

Which do you prefer when you press the "Load more" button?

We'll support both, this is about what the *default* should be.

1. Jump to the newest post (current behaviour)
2. Stay at the oldest post
```

Use the result of that to determine the default.

Reply:
"There are behavioral inconsistencies" is not the same as "the default is secretly the opposite of what you think".

I agree that when the user *actively* refreshes, the timeline should move to the top (in the *default* newest-first order).

There's no point in having a poll about the load more behavior, because it should be determined by the reading order, where it doesn't "jump" in either case, but remains at the user's last-read (from the *default* top or optional bottom) post.

Reply:
> "There are behavioral inconsistencies" is not the same as "the default is secretly the opposite of what you think".

That is true. 

But that's not what I'm saying.

I'm saying that the default behaviour is inconsistent and is a source (probably a significant one) of user confusion and dissatisfaction with the app.

By pointing out the inconsistency, I intend to suggest that when you say "I believe that today's behavior is correct for this use case" I think you may not have considered all the ways that the app behaves.

In response, you suggested "I agree that when the user actively refreshes, the timeline should move to the top (in the default newest-first order)."

That is something that we could do (not in this PR, and I'm not suggesting we actually do it, I'm using "could" to mean only "This is a technically possible change").

But you've also said (about default reading order) "we should absolutely not change such a fundamental default behavior on a whim or due to selection bias."

I agree with that too -- that's why I suggested the poll, so that it's not a whim, or selection bias.

But I feel this is also inconsistent. On one hand you are suggesting that we should *not* change fundamental default behaviour like this, on the other hand you suggest we should change the user's reading position when they swipe-to-refresh, which I think would definitely be a change to fundamental behaviour.

I think that may be why some of this discussion is challenging; it feels like you are arguing two different and opposite positions, depending on the behaviour that you want to see, and it's difficult to address your concerns because of that.

---

Aside: As I've already said, I agree with when you say "absolutely not change such a fundamental default behaviour on a whim or due to selection bias". 

And I absolutely do *not* want to discuss the specific decision here and derail the conversation, but I'll note that has absolutely happened in the past, probably most recently with https://github.com/tuskyapp/Tusky/issues/2552. 

---

Finally:

> There's no point in having a poll about the load more behavior, because it should be determined by the reading order, where it doesn't "jump" in either case, but remains at the user's last-read (from the default top or optional bottom) post.

That's exactly what this PR does (once I revert https://github.com/tuskyapp/Tusky/commit/1480c6aa3ac5c0c2d362fb271f47ea2259ab14e2, see my other reply https://github.com/tuskyapp/Tusky/pull/3000#issuecomment-1369733104 about that).

The only thing the poll is for is decide the default value for the preference, `OLDEST_FIRST` or `NEWEST_FIRST`.